sdks/go/pkg/beam/core/runtime/graphx/cobgk.go

```diff
@@ -0,0 +1,107 @@
+// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements.  See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package graphx
+
+import (
+	"fmt"
+
+	"github.com/apache/beam/sdks/go/pkg/beam/core/graph"
+	"github.com/apache/beam/sdks/go/pkg/beam/core/graph/coder"
+	"github.com/apache/beam/sdks/go/pkg/beam/core/typex"
+	"github.com/apache/beam/sdks/go/pkg/beam/core/util/reflectx"
+)
+
+// CoGBK support
+//
+// TODO(BEAM-490): CoGBK is not a supported model primitive, so similarly to other
+// SDKs, a translation into GBK is performed to run on portable runners. Due to
+// various constraints and to preserve CoGBK as a first-class concept in areas
+// such as type-checking and for non-portability runners, the Go translation is
+// done by the framework instead of user code.
+//
+// The basic translation approach is as follows:
+//
+//         In1 : KV<K,X>        In2 : KV<K,Y>
+//               |                   |
+//               ------ CoGBK --------
+//                        |
+//                Out: CoGBK<K,X,Y>
+//
+// is expanded to a flattened raw union value ("RUV") typed as KV<int,[]byte>:
+//
+//         In1 : KV<K,X>        In2 : KV<K,Y>
```

Comment:
This is an excellent explanation of what's going on. Nicely done.

Reply:
Please fix the filename. Almost lost my mind not being able to find this file in my snapshot. ;-)